File: packages/on_audio_query/on_audio_query_ios/ios/Classes/methods/queries/ArtworkQuery.swift
import Flutter
import MediaPlayer

class ArtworkQuery {
    
    private let pluginDir = "/on_audio_query/mediastore/artworks/"
    
    private var args: [String: Any]
    private var result: FlutterResult
    
    init(call: FlutterMethodCall, result: @escaping FlutterResult) {
        // To make life easy, add all arguments inside a map.
        self.args = call.arguments as! [String: Any]
        self.result = result
    }
    
    // [IOS] has a different artwork system and you can "query" using normal "queryAudios, .."
    // [Android] can't "query" artwork at the same time as "queryAudios", so we need to "query"
    // using a different method(queryArtwork).
    //
    // To match both [IOS] and [Android], [queryArtwork] is the only way to get artwork.
    //
    // Not the best solution but, at least here we can select differents formats and size.
    func queryArtwork() {
        // None of this arguments can be null.
        // The id of the [Audio] or [Album].
        let id = args["id"] as! Int
        
        // The uri [0]: Audio and [1]: Album.
        let uri = args["type"] as! Int
        
        // (To match android side, let's call "cursor").
        var cursor: MPMediaQuery?
        var filter: MPMediaPropertyPredicate?
        
        // If [uri] is 0: artwork from [Audio]
        // If [uri] is 1: artwork from [Album]
        // If [uri] is 2: artwork from [Playlist]
        // If [uri] is 3: artwork from [Artist]
        switch uri {
        case 0:
            filter = MPMediaPropertyPredicate.init(value: id, forProperty: MPMediaItemPropertyPersistentID)
            cursor = MPMediaQuery.songs()
        case 1:
            filter = MPMediaPropertyPredicate.init(value: id, forProperty: MPMediaItemPropertyAlbumPersistentID)
            cursor = MPMediaQuery.albums()
        case 2:
            filter = MPMediaPropertyPredicate.init(value: id, forProperty: MPMediaPlaylistPropertyPersistentID)
            cursor = MPMediaQuery.playlists()
        case 3:
            filter = MPMediaPropertyPredicate.init(value: id, forProperty: MPMediaItemPropertyArtistPersistentID)
            cursor = MPMediaQuery.artists()
        case 4:
            filter = MPMediaPropertyPredicate.init(value: id, forProperty: MPMediaItemPropertyGenrePersistentID)
            cursor = MPMediaQuery.genres()
        default:
            filter = nil
            cursor = nil
        }
        
        // Request permission status from the 'main' method.
        let hasPermission = SwiftOnAudioQueryPlugin().checkPermission()
        
        // We cannot 'query' without permission so, throw a PlatformException.
        if !hasPermission {
            // Method from 'MethodChannel' (method)
            self.result(
                FlutterError.init(
                    code: "403",
                    message: "The app doesn't have permission to read files.",
                    details: "Call the [permissionsRequest] method or install a external plugin to handle the app permission."
                )
            )
            
            // 'Exit' the function
            return
        }
        
        //
        if cursor != nil && filter != nil {
            cursor?.addFilterPredicate(filter!)
            
            // This filter will avoid audios/songs outside phone library(cloud).
            let cloudFilter = MPMediaPropertyPredicate.init(
                value: false,
                forProperty: MPMediaItemPropertyIsCloudItem
            )
            cursor?.addFilterPredicate(cloudFilter)
            
            // Query everything in background for a better performance.
            loadArtwork(cursor: cursor, id: id, uri: uri)
        } else {
            // Return to Flutter
            result(nil)
        }
    }
    
    private func loadArtwork(cursor: MPMediaQuery!, id: Int, uri: Int) {
        // The size of the image.
        let size = args["size"] as! Int
        
        // The size of the image.
        var quality = args["quality"] as! Int
        if (quality > 100) {
            quality = 50
        }
        
        // The format [JPEG] or [PNG].
        let format = args["format"] as! Int
        
        DispatchQueue.global(qos: .userInitiated).async {
            var artwork: Data?
            var item: MPMediaItem?
            let fixedQuality = CGFloat(Double(quality) / 100.0)
            
            // If [uri] is 0: artwork is from [Audio]
            // If [uri] is 1, 2 or 3: artwork is from [Album], [Playlist] or [Artist]
            if uri == 0 {
                // Since all id are unique, we can safely call the first item.
                item = cursor!.items?.first
            } else {
                // Since all id are unique, we can safely call the first item.
                item = cursor!.collections?.first?.items.first
            }
            
            // If [format] is 0: will be [JPEG]
            // If [format] is 1: will be [PNG]
            if format == 0 {
                artwork = item?.artwork?.image(at: CGSize(width: size, height: size))?.jpegData(compressionQuality: fixedQuality)
            } else {
                // [PNG] format will return a high image quality.
                artwork = item?.artwork?.image(at: CGSize(width: size, height: size))?.pngData()
            }
            
            // After finish the "query", go back to the "main" thread(You can only call flutter
            // inside the main thread).
            DispatchQueue.main.async {
                // We don't need a "empty" image so, return null to avoid problems.
                if artwork != nil && artwork!.isEmpty {
                    artwork = nil
                }
                
                // Define the value to cache or no the artwork inside the app dir.
                let cacheArtwork = self.args["cacheArtwork"] as! Bool
                
                // This parameter will only be defined if [cacheArtwork] is true and
                // the artwork can be cached.
                var artPath: String? = nil
                
                // Check the value of artwork and if we need to 'cache'.
                if artwork != nil && cacheArtwork {
                    artPath = self.cacheArtwork(format: format, artwork: artwork!)
                }
                
                // Send the result.
                self.result(
                    [
                        "_id": id,
                        "artwork": artwork as Any,
                        "path": artPath as Any,
                        "type": format == 0 ? "JPEG" : "PNG"
                    ]
                )
            }
        }
    }
    
    private func cacheArtwork(format: Int, artwork: Data) -> String? {
        // The file type.
        var fileType: String
        
        // Check the file type.
        if format == 0 {
            fileType = ".jpeg"
        } else {
            fileType = ".png"
        }
        
        // All artworks will be 'cached' with audio/album/artist/playlist
        // or genre [id] and the artwork [type].
        let fileName = String((args["id"] as! Int)) + fileType
        
        // Define if the artwork will be saved in a tmp folder.
        let cacheTemporarily = args["cacheTemporarily"] as! Bool
        let dirType: FileManager.SearchPathDirectory
        
        if cacheTemporarily {
            dirType = FileManager.SearchPathDirectory.cachesDirectory
        } else {
            dirType = FileManager.SearchPathDirectory.documentDirectory
        }
        
        // Define if the file will be overwritten.
        let overrideCache = args["overrideCache"] as! Bool
        
        // Check if the specified directory exists.
        if let dir: String = NSSearchPathForDirectoriesInDomains(
            dirType,
            .userDomainMask,
            true
        ).first {
            // Join the specific app dir with the plugin dir and check if
            // the directory exists.
            if var pathUrl = URL(string: dir + pluginDir) {
                
                // Check if the artwork already exist.
                let fileExists = FileManager.default.fileExists(
                    atPath: pathUrl.absoluteString + fileName
                )
                
                // If [overrideCache] is true the artwork will be overwritten.
                //
                // If [overrideCache] is false and the file already exists, return
                // the path to this artwork.
                if !overrideCache && fileExists {
                    return pathUrl.absoluteString
                }
                
                // Create the plugin directory if is none.
                if !FileManager.default.fileExists(atPath: pathUrl.absoluteString) {
                    do {
                        try FileManager.default.createDirectory(
                            atPath: pathUrl.absoluteString,
                            withIntermediateDirectories: true,
                            attributes: nil
                        )
                    } catch {
                        print("ArtworkQuery::cacheArtwork(fileExists) -> \(error)")
                    }
                }
                
                // Now the the file dir and define a new file(artwork).
                pathUrl.appendPathComponent(fileName)
                
                // Add a 'file:' schema to url.
                let fixedPath = URL(fileURLWithPath: pathUrl.absoluteString)
                
                // Try write the artwork.
                do {
                    // Try to write the file.
                    try artwork.write(to: fixedPath, options: Data.WritingOptions.atomic)
                    
                    // Return the path to this artwork.
                    return pathUrl.absoluteString
                } catch {
                    print("ArtworkQuery::cacheArtwork(write) -> \(error)")
                }
            }
        }
        
        // If anything goes wrong, return nil.
        return nil
    }
}

File: packages/on_audio_query/on_audio_query_android/android/src/main/kotlin/com/lucasjosino/on_audio_query/methods/queries/ArtworkQuery.kt
package com.lucasjosino.on_audio_query.methods.queries

import android.content.ContentResolver
import android.content.ContentUris
import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.media.MediaMetadataRetriever
import android.net.Uri
import android.os.Build
import android.util.Size
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.lucasjosino.on_audio_query.controllers.PermissionController
import com.lucasjosino.on_audio_query.methods.helper.QueryHelper
import com.lucasjosino.on_audio_query.types.checkArtworkFormat
import com.lucasjosino.on_audio_query.types.checkArtworkType
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.ByteArrayOutputStream
import java.io.FileInputStream

/** ArtworkQuery */
class ArtworkQuery : ViewModel() {

    //Main parameters
    private val helper = QueryHelper()
    private var type: Int = -1
    private var id: Number = 0
    private var quality: Int = 50
    private var size: Int = 100

    //
    private var query: Uri = Uri.EMPTY

    // None of this methods can be null.
    private lateinit var uri: Uri
    private lateinit var resolver: ContentResolver
    private lateinit var format: Bitmap.CompressFormat

    /**
     * Method to "query" all albums.
     *
     * Parameters:
     *   * [context]
     *   * [result]
     *   * [call]
     */
    fun queryArtwork(context: Context, result: MethodChannel.Result, call: MethodCall) {
        resolver = context.contentResolver

        // The [id] of the audio/album.
        id = call.argument<Number>("id")!!

        // If the [size] is null, will be [100].
        size = call.argument<Int>("size")!!

        // Define the quality of image.
        // The [quality] value cannot be greater than 100 so, we check and if is, set to [50].
        quality = call.argument<Int>("quality")!!
        if (quality > 100) quality = 50

        // Check format:
        //   * [0]: JPEG
        //   * [1]: PNG
        format = checkArtworkFormat(call.argument<Int>("format")!!)

        // Check uri:
        //   * [0]: Audio.
        //   * [1]: Album.
        //   * [2]: Playlist.
        //   * [3]: Artist.
        //   * [4]: Genre.
        uri = checkArtworkType(call.argument<Int>("type")!!)

        // Define the [type]:
        type = call.argument<Int>("type")!!

        // Request permission status from the 'main' method.
        val hasPermission: Boolean = PermissionController().permissionStatus(context)

        // We cannot 'query' without permission so, throw a PlatformException.
        if (!hasPermission) {
            // Method from 'MethodChannel' (method)
            result.error(
                "403",
                "The app doesn't have permission to read files.",
                "Call the [permissionsRequest] method or install a external plugin to handle the app permission."
            )

            // 'Exit' the function
            return
        }

        // Query everything in background for a better performance.
        viewModelScope.launch {
            // Start 'querying'.
            var resultArtwork: ByteArray? = loadArt()

            // Sometimes android will extract a 'wrong' or 'empty' artwork. Just set as null.
            if (resultArtwork != null && resultArtwork.isEmpty()) {
                resultArtwork = null
            }

            // After loading the information, send the 'result'.
            result.success(
                hashMapOf<String, Any?>(
                    "_id" to id.toInt(),
                    "artwork" to resultArtwork,
                    "path" to query.path,
                    "type" to format.name
                )
            )
        }
    }

    //Loading in Background
    @Suppress("BlockingMethodInNonBlockingContext")
    private suspend fun loadArt(): ByteArray? = withContext(Dispatchers.IO) {
        // Empty array.
        var artData: ByteArray? = null

        // In this case we need check the [Android] version and query [type].
        //
        // If [Android] >= 29/Q:
        //   * We have a limited access to files/folders and we use [loadThumbnail].
        // If [Android] < 29/Q:
        //   * We use the [embeddedPicture] from [MediaMetadataRetriever] to get the image.
        if (Build.VERSION.SDK_INT >= 29) {
            // Try / Catch to avoid problems.
            try {
                // If [type] is 2, 3 or 4, we need to 'get' the first item from playlist or artist.
                // We'll use the first artist audio to 'simulate' the artwork.
                //
                // Type:
                //   * [2]: Playlist.
                //   * [3]: Artist.
                //   * [4]: Genre.
                //
                // Due old problems with [MethodChannel] the [id] is defined as [Number].
                // Here we convert to [Long]
                query = if (type == 2 || type == 3 || type == 4) {
                    val item = helper.loadFirstItem(type, id, resolver) ?: return@withContext null
                    ContentUris.withAppendedId(uri, item.toLong())
                } else {
                    ContentUris.withAppendedId(uri, id.toLong())
                }

                val bitmap = resolver.loadThumbnail(query, Size(size, size), null)
                artData = convertOrResize(bitmap = bitmap)!!
            } catch (e: Exception) {
                // Some problem can occur, we hide to not "flood" the terminal.
//                Log.i("on_audio_error: ", e.toString())
            }
        } else {
            // If [uri == Audio]:
            //   * Load the first [item] from cursor using the [id] as filter.
            // else:
            //   * Load the first [item] from [album] using the [id] as filter.
            //
            // If [item] return null, no audio/album has found, just return null.
            val item = helper.loadFirstItem(type, id, resolver) ?: return@withContext null
            try {
                // I tried both [_data] and [_uri], none of them work.
                // So we use the [_data] inside the [FileInputStream] and take the
                // [fd(FileDescriptor)].
                val file = FileInputStream(item)
                val metadata = MediaMetadataRetriever()

                // Most of the cases the error occurred here.
                metadata.setDataSource(file.fd)
                val image = metadata.embeddedPicture

                // Check if [image] null.
                artData = convertOrResize(byteArray = image) ?: return@withContext null

                // [close] can only be called using [Android] >= 29/Q.
                if (Build.VERSION.SDK_INT >= 29) metadata.close()
            } catch (e: Exception) {
                // Some problem can occur, we hide to not "flood" the terminal.
//                Log.i("on_audio_error: ", e.toString())
            }
        }
        // After finish the "query", go back to the "main" thread(You can only call flutter
        // inside the main thread).
        return@withContext artData
    }

    //
    private fun convertOrResize(bitmap: Bitmap? = null, byteArray: ByteArray? = null): ByteArray? {
        val convertedBytes: ByteArray?
        val byteArrayBase = ByteArrayOutputStream()
        try {
            // If [bitmap] isn't null:
            //   * The image(bitmap) is from first method. (Android >= 29/Q).
            // else:
            //   * The image(bytearray) is from second method. (Android < 29/Q).
            if (bitmap != null) {
                bitmap.compress(format, quality, byteArrayBase)
            } else {
                val convertedBitmap = BitmapFactory.decodeByteArray(byteArray, 0, byteArray!!.size)
                convertedBitmap.compress(format, quality, byteArrayBase)
            }
        } catch (e: Exception) {
            //Log.i("Error", e.toString())
        }
        convertedBytes = byteArrayBase.toByteArray()
        byteArrayBase.close()
        return convertedBytes
    }
}